File: Daily/2023/547_numberOfProvinces.cs
public class UnionFind
{

    public int[] parents;
    public int[] ranks;

    public UnionFind(int n)
    {
        parents = new int[n + 1];
        for (int idx = 1; idx < n + 1; idx++)
        {
            parents[idx] = idx;
        }
        ranks = new int[n + 1];
        for (int idx = 1; idx < n + 1; idx++)
        {
            ranks[idx] = 1;
        }
    }

    public int Find(int n)
    {
        int root = n;
        while (root != parents[root])
        {
            root = parents[root];
        }

        int nxt = -1;
        int cur = n;
        while (cur != root)
        {
            nxt = parents[cur];
            parents[cur] = root;
            cur = nxt;
        }

        return root;
    }

    public void Union(int n1, int n2)
    {
        int r1 = Find(n1), r2 = Find(n2);
        if (r1 == r2)
        {
            return;
        }
        if (ranks[r1] > ranks[r2])
        {
            parents[r2] = r1;
            ranks[r1] += ranks[r2];
        }
        else
        {
            parents[r1] = r2;
            ranks[r2] += ranks[r1];
        }
    }
}

public class Solution
{
    public int FindCircleNum(int[][] isConnected)
    {
        int n = isConnected.Length;
        var unionFind = new UnionFind(n);

        for (int idx = 1; idx < n + 1; idx++)
        {
            for (int jdx = idx + 1; jdx < n + 1; jdx++)
            {
                if (idx == jdx) continue;
                if (isConnected[idx - 1][jdx - 1] == 1)
                {
                    unionFind.Union(idx, jdx);
                }
            }
        }

        var vis = new HashSet<int>();
        for (int idx = 1; idx < n + 1; idx++)
        {
            vis.Add(unionFind.Find(idx));
        }

        return vis.Count();
    }
}

File: Daily/2023/547_numberOfProvinces.py
from typing import *


class UnionFind:
    def __init__(self, n):
        self.parents = [idx for idx in range(n + 1)]
        self.ranks = [1 for _ in range(n + 1)]

    def find(self, n):
        root = n
        while root != self.parents[root]:
            root = self.parents[root]

        cur = n
        while cur != root:
            nxt = self.parents[cur]
            self.parents[cur] = root
            cur = nxt

        return root

    def union(self, n1, n2):
        r1, r2 = self.find(n1), self.find(n2)
        if r1 == r2:
            return

        if self.ranks[r1] > self.ranks[r2]:
            self.parents[r2] = r1
            self.ranks[r1] += self.ranks[r2]
        else:
            self.parents[r1] = r2
            self.ranks[r2] += self.ranks[r1]


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        union_find = UnionFind(n)

        for idx in range(1, n + 1):
            for jdx in range(idx + 1, n + 1):
                if idx == jdx:
                    continue
                if isConnected[idx - 1][jdx - 1]:
                    union_find.union(idx, jdx)

        vis = set()
        for idx in range(1, n + 1):
            vis.add(union_find.find(idx))

        return len(vis)
